Limit request line length (#784)

cesarbs opened the pull request:
#784

In order to address this, I've added new `MemoryPoolIterator.Seek` overloads. One allows setting a limit on how many bytes can be scanned, and returns (via an `out` parameter) the number of bytes scanned; the other one allows passing in another `MemoryPoolIterator` to server as a limit on the seek operation.

This change changes the behavior of `Frame.TakeStartLine` to wait until the final LF has been received before parsing the line. If LF is not found before the line limit, the request is rejected. Because of this change, `BadHttpRequestTests.ServerClosesConnectionAsSoonAsBadRequestLineIsDetected` is not a meaningful test anymore.

cc @halter73 @mikeharder @davidfowl

Tratcher commented on src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerOptions.cs:
```diff
@@ -57,13 +60,35 @@ public class KestrelServerOptions
+            }
+            set
+            {
+                if (value <= 0)
```
What's the shortest practical request line? `GET / HTTP/1.1`? Any reason to allow the limit to be lower than 16?

halter73 replied:
Any rule saying a method has to be at least 3 characters (even if they always are in practice)? You could argue there is no reason for even 16 to be valid. I don't see any reason to be too smart with this limit.

Tratcher commented on test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/MaxRequestLineSizeTests.cs:
```diff
@@ -0,0 +1,66 @@
+        [Theory]
+        [InlineData(1, 2)]
+        [InlineData(int.MaxValue - 1, int.MaxValue)]
+        public void ServerFailsToStartWhenMaxRequestLineSizeIsLessThanMaxRequestBufferSize(long maxRequestBufferSize, int maxRequestLineSize)
```
Isn't this description backwards?

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs:
```diff
@@ -277,6 +328,16 @@ public unsafe int Seek(ref Vector<byte> byte0Vector)
+                        bytesScanned++;
+
+                        if ((bytesScanned > limitBytes) ||
+                            (!limitIterator.IsDefault && block == limitIterator.Block && index > limitIterator.Index))
```
instead of doing this complicated comparison for every byte, couldn't you adjust `pEnd` to match the limit index?

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs:
```diff
@@ -261,11 +293,30 @@ public unsafe int Seek(ref Vector<byte> byte0Vector)
+                                bytesScanned += _vectorSpan;
+
+                                if ((bytesScanned > limitBytes) ||
+                                    (!limitIterator.IsDefault && block == limitIterator.Block && index > limitIterator.Index))
```
I think all these `!limitIterator.IsDefault` checks are unnecessary considering `block` will never be null.

davidfowl commented on src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerOptions.cs:
```diff
@@ -57,13 +60,35 @@ public class KestrelServerOptions
+        /// <remarks>
+        /// Defaults to 8,192 bytes (8 KB).
+        /// </remarks>
+        public int MaxRequestLineSize
```
I think we should move these settings to a sub object so the top level class isn;t polluted.

halter73 replied:
"Limits" maybe? For both max lengths and timeouts.

davidfowl replied:
Yep

cesarbs commented:
Updated:

* Removed unnecessary `IsDefault` checks
* Set `pEnd` based on limit instead of checking for limit on each iteration of the `do while` loops
* Fixed inconsistencies with the iterator state when `Seek()` fails. It's now always at the expected position: one byte past the byte limit in seeks with byte limits, and exactly at the limit iterator in seeks with an iterator limit.

cesarbs commented:
Also, I removed the private overload for `Seek()` which took both a byte and an iterator limit and tried to handle those together. There's now a separate `Seek()` for each option. I'm thinking of reducing the resulting code duplication by adding a `checkLimit` callback, but I need to make sure that doesn't negatively impact perf.

cesarbs commented:
Another update: moves `MaxRequestLineSize` to the new class `KestrelServerLimits`, accessible via the `Limits` property of `KestrelServerOptions`.

cesarbs commented:
Rebased.

cesarbs commented:
@davidfowl I'm thinking of moving `MaxRequestBufferSize` to the new `KestrelServerLimits` class and marking the one in `KestrelServerOptions` as deprecated, and keeping it as a proxy to the new one so we don't break anything. What do you think?

davidfowl commented:
We should talk to @Eilon about using `[Obsolete]` to deprecate APIs.

davidfowl commented on test/Microsoft.AspNetCore.Server.KestrelTests/MemoryPoolIteratorTests.cs:
```diff
@@ -406,6 +411,223 @@ public void KnownMethodsAreInterned(string input, string expected)
+            // Arrange
+            var seekVector = new Vector<byte>((byte)seek);
+
+            var block = _pool.Lease();
```
put the lease return in a try finally

cesarbs replied:
Ugh :( Ok.

Eilon commented:
@cesarbs you can follow the guidelines here to mark things as obsolete: https://github.com/aspnet/Home/wiki/Engineering-guidelines#breaking-changes

Then log a new bug to remove the old property and @muratg can triage into the 2.0.0 milestone and mark as breaking-change.

cesarbs commented:
@davidfowl Put tests in try/finally.

@Eilon @davidfowl Moved property and marked old one as obsolete. Please review the last commit to check that I did everything right.

muratg commented:
https://github.com/aspnet/KestrelHttpServer/issues/1027 to track the future removal of it

Eilon commented on src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerOptions.cs:
```diff
@@ -41,29 +34,37 @@ public class KestrelServerOptions
+        /// When set to null, the size of the request buffer is unlimited.
         /// Defaults to 1,048,576 bytes (1 MB).
         /// </remarks>
+        [Obsolete]
```
Please follow the guidelines: Include the obsoletion message here.

cesarbs replied:
https://github.com/aspnet/KestrelHttpServer/pull/1029

benaadams commented on src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs:
```diff
@@ -259,13 +379,33 @@ public unsafe int Seek(ref Vector<byte> byte0Vector)
-                            _index = index + FindFirstEqualByte(ref byte0Equals);
+
+                            var firstEqualByteIndex = FindFirstEqualByte(ref byte0Equals);
+                            var vectorBytesScanned = firstEqualByteIndex + 1;
```
@cesarbs The vectorBytesScanned variable isn't used here
